File: skills/runninghub-api-dev/assets/sdk/python/runninghub_client.py
#!/usr/bin/env python3
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import requests


DEFAULT_BASE_URL = "https://www.runninghub.cn"


@dataclass
class RunningHubApiError(Exception):
    """统一封装 RunningHub API 的错误。"""

    message: str
    response: dict[str, Any] | None = None
    http_status: int | None = None

    def __str__(self) -> str:
        detail = self.message
        if self.http_status is not None:
            detail = f"HTTP {self.http_status}: {detail}"
        if self.response:
            code = self.response.get("code")
            msg = self.response.get("msg") or self.response.get("message")
            if code is not None or msg is not None:
                detail = f"{detail} | code={code} msg={msg}"
        return detail


def mask_api_key(api_key: str, visible: int = 4) -> str:
    if len(api_key) <= visible * 2:
        return "*" * len(api_key)
    return f"{api_key[:visible]}{'*' * (len(api_key) - visible * 2)}{api_key[-visible:]}"


class RunningHubClient:
    """RunningHub 底层 HTTP 客户端。"""

    def __init__(
        self,
        api_key: str,
        base_url: str = DEFAULT_BASE_URL,
        timeout: int = 60,
        session: requests.Session | None = None,
    ) -> None:
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = session or requests.Session()

    def _headers(
        self,
        *,
        include_auth: bool = True,
        include_json: bool = True,
        extra_headers: dict[str, str] | None = None,
    ) -> dict[str, str]:
        headers = {"Host": "www.runninghub.cn"}
        if include_auth:
            headers["Authorization"] = f"Bearer {self.api_key}"
        if include_json:
            headers["Content-Type"] = "application/json"
        if extra_headers:
            headers.update(extra_headers)
        return headers

    def _request_json(
        self,
        method: str,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        json_body: dict[str, Any] | None = None,
        include_auth: bool = True,
        include_json: bool = True,
        extra_headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        response = self.session.request(
            method=method.upper(),
            url=f"{self.base_url}{path}",
            params=params,
            json=json_body,
            headers=self._headers(
                include_auth=include_auth,
                include_json=include_json,
                extra_headers=extra_headers,
            ),
            timeout=self.timeout,
        )
        try:
            payload = response.json()
        except requests.JSONDecodeError as error:
            raise RunningHubApiError(
                message=f"接口未返回 JSON：{response.text[:200]}",
                http_status=response.status_code,
            ) from error

        if response.status_code >= 400:
            raise RunningHubApiError(
                message="HTTP 请求失败",
                response=payload,
                http_status=response.status_code,
            )
        return payload

    def check_account(
        self,
        *,
        include_auth: bool = True,
        include_body: bool = True,
    ) -> dict[str, Any]:
        body = {"apikey": self.api_key} if include_body else {}
        return self._request_json(
            "POST",
            "/uc/openapi/accountStatus",
            json_body=body,
            include_auth=include_auth,
        )

    def get_ai_app_demo(
        self,
        webapp_id: str | int,
        *,
        include_auth: bool = True,
    ) -> dict[str, Any]:
        return self._request_json(
            "GET",
            "/api/webapp/apiCallDemo",
            params={"apiKey": self.api_key, "webappId": str(webapp_id)},
            json_body=None,
            include_auth=include_auth,
            include_json=False,
        )

    def upload_file(self, file_path: str | Path) -> dict[str, Any]:
        target = Path(file_path)
        with target.open("rb") as file_obj:
            response = self.session.post(
                f"{self.base_url}/openapi/v2/media/upload/binary",
                headers=self._headers(include_json=False),
                files={"file": (target.name, file_obj)},
                timeout=self.timeout,
            )
        try:
            payload = response.json()
        except requests.JSONDecodeError as error:
            raise RunningHubApiError(
                message=f"上传接口未返回 JSON：{response.text[:200]}",
                http_status=response.status_code,
            ) from error
        if response.status_code >= 400:
            raise RunningHubApiError(
                message="上传接口 HTTP 失败",
                response=payload,
                http_status=response.status_code,
            )
        return payload

    def run_ai_app(
        self,
        *,
        webapp_id: str | int,
        node_info_list: list[dict[str, Any]],
        webhook_url: str | None = None,
        instance_type: str | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "webappId": str(webapp_id),
            "apiKey": self.api_key,
            "nodeInfoList": node_info_list,
        }
        if webhook_url:
            payload["webhookUrl"] = webhook_url
        if instance_type:
            payload["instanceType"] = instance_type
        return self._request_json("POST", "/task/openapi/ai-app/run", json_body=payload)

    def query_status(
        self,
        task_id: str | int,
        *,
        include_auth: bool = True,
        include_body_api_key: bool = True,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {"taskId": str(task_id)}
        if include_body_api_key:
            payload["apiKey"] = self.api_key
        return self._request_json(
            "POST",
            "/task/openapi/status",
            json_body=payload,
            include_auth=include_auth,
        )

    def query_outputs(
        self,
        task_id: str | int,
        *,
        include_auth: bool = True,
        include_body_api_key: bool = True,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {"taskId": str(task_id)}
        if include_body_api_key:
            payload["apiKey"] = self.api_key
        return self._request_json(
            "POST",
            "/task/openapi/outputs",
            json_body=payload,
            include_auth=include_auth,
        )

    def query_v2(
        self,
        task_id: str | int,
        *,
        include_auth: bool = True,
    ) -> dict[str, Any]:
        return self._request_json(
            "POST",
            "/openapi/v2/query",
            json_body={"taskId": str(task_id)},
            include_auth=include_auth,
        )

    def save_json(self, path: str | Path, payload: dict[str, Any]) -> None:
        target = Path(path)
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

File: skills/runninghub-api-dev/assets/sdk/node/runninghub-client.mjs
import { readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';

export const DEFAULT_BASE_URL = 'https://www.runninghub.cn';

export class RunningHubApiError extends Error {
  constructor(message, { response = null, httpStatus = null } = {}) {
    const details = [];
    if (httpStatus !== null && httpStatus !== undefined) {
      details.push(`HTTP ${httpStatus}`);
    }
    if (response && typeof response === 'object') {
      const code = response.code;
      const msg = response.msg ?? response.message;
      if (code !== undefined || msg !== undefined) {
        details.push(`code=${code} msg=${msg}`);
      }
    }
    super(details.length ? `${message} | ${details.join(' ')}` : message);
    this.name = 'RunningHubApiError';
    this.response = response;
    this.httpStatus = httpStatus;
  }
}

export function maskApiKey(apiKey, visible = 4) {
  if (apiKey.length <= visible * 2) {
    return '*'.repeat(apiKey.length);
  }
  return `${apiKey.slice(0, visible)}${'*'.repeat(apiKey.length - visible * 2)}${apiKey.slice(-visible)}`;
}

export class RunningHubClient {
  constructor({
    apiKey,
    baseUrl = DEFAULT_BASE_URL,
    timeoutMs = 60_000,
    fetchImpl = globalThis.fetch,
  }) {
    if (!apiKey) {
      throw new Error('缺少 apiKey');
    }
    if (typeof fetchImpl !== 'function') {
      throw new Error('当前环境缺少 fetch 实现');
    }
    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/+$/u, '');
    this.timeoutMs = timeoutMs;
    this.fetchImpl = fetchImpl;
  }

  buildHeaders({
    includeAuth = true,
    includeJson = true,
    extraHeaders = {},
  } = {}) {
    const headers = {
      Host: 'www.runninghub.cn',
      ...extraHeaders,
    };
    if (includeAuth) {
      headers.Authorization = `Bearer ${this.apiKey}`;
    }
    if (includeJson) {
      headers['Content-Type'] = 'application/json';
    }
    return headers;
  }

  buildUrl(pathname, params = null) {
    const url = new URL(`${this.baseUrl}${pathname}`);
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined && value !== null) {
          url.searchParams.set(key, String(value));
        }
      }
    }
    return url.toString();
  }

  async requestJson(
    method,
    pathname,
    {
      params = null,
      jsonBody = undefined,
      body = undefined,
      includeAuth = true,
      includeJson = true,
      extraHeaders = {},
    } = {},
  ) {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetchImpl(this.buildUrl(pathname, params), {
        method: method.toUpperCase(),
        headers: this.buildHeaders({ includeAuth, includeJson, extraHeaders }),
        body: body ?? (jsonBody === undefined ? undefined : JSON.stringify(jsonBody)),
        signal: controller.signal,
      });
      const rawText = await response.text();
      let payload;
      try {
        payload = rawText ? JSON.parse(rawText) : {};
      } catch (error) {
        throw new RunningHubApiError(`接口未返回 JSON：${rawText.slice(0, 200)}`, {
          httpStatus: response.status,
        });
      }
      if (response.status >= 400) {
        throw new RunningHubApiError('HTTP 请求失败', {
          response: payload,
          httpStatus: response.status,
        });
      }
      return payload;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async checkAccount({ includeAuth = true, includeBody = true } = {}) {
    const jsonBody = includeBody ? { apikey: this.apiKey } : {};
    return this.requestJson('POST', '/uc/openapi/accountStatus', {
      jsonBody,
      includeAuth,
    });
  }

  async getAiAppDemo(webappId, { includeAuth = true } = {}) {
    return this.requestJson('GET', '/api/webapp/apiCallDemo', {
      params: {
        apiKey: this.apiKey,
        webappId: String(webappId),
      },
      includeAuth,
      includeJson: false,
    });
  }

  async uploadFile(filePath) {
    const fileBuffer = await readFile(filePath);
    const fileName = path.basename(filePath);
    const formData = new FormData();
    formData.set('file', new Blob([fileBuffer]), fileName);

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetchImpl(
        `${this.baseUrl}/openapi/v2/media/upload/binary`,
        {
          method: 'POST',
          headers: this.buildHeaders({ includeJson: false }),
          body: formData,
          signal: controller.signal,
        },
      );
      const rawText = await response.text();
      let payload;
      try {
        payload = rawText ? JSON.parse(rawText) : {};
      } catch (error) {
        throw new RunningHubApiError(`上传接口未返回 JSON：${rawText.slice(0, 200)}`, {
          httpStatus: response.status,
        });
      }
      if (response.status >= 400) {
        throw new RunningHubApiError('上传接口 HTTP 失败', {
          response: payload,
          httpStatus: response.status,
        });
      }
      return payload;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async runAiApp({
    webappId,
    nodeInfoList,
    webhookUrl = undefined,
    instanceType = undefined,
  }) {
    const payload = {
      webappId: String(webappId),
      apiKey: this.apiKey,
      nodeInfoList,
    };
    if (webhookUrl) {
      payload.webhookUrl = webhookUrl;
    }
    if (instanceType) {
      payload.instanceType = instanceType;
    }
    return this.requestJson('POST', '/task/openapi/ai-app/run', {
      jsonBody: payload,
    });
  }

  async queryStatus(taskId, { includeAuth = true, includeBodyApiKey = true } = {}) {
    const payload = { taskId: String(taskId) };
    if (includeBodyApiKey) {
      payload.apiKey = this.apiKey;
    }
    return this.requestJson('POST', '/task/openapi/status', {
      jsonBody: payload,
      includeAuth,
    });
  }

  async queryOutputs(taskId, { includeAuth = true, includeBodyApiKey = true } = {}) {
    const payload = { taskId: String(taskId) };
    if (includeBodyApiKey) {
      payload.apiKey = this.apiKey;
    }
    return this.requestJson('POST', '/task/openapi/outputs', {
      jsonBody: payload,
      includeAuth,
    });
  }

  async queryV2(taskId, { includeAuth = true } = {}) {
    return this.requestJson('POST', '/openapi/v2/query', {
      jsonBody: { taskId: String(taskId) },
      includeAuth,
    });
  }

  async saveJson(targetPath, payload) {
    await writeFile(targetPath, `${JSON.stringify(payload, null, 2)}\n`, 'utf8');
  }
}
